test/test_torch.py

```diff
@@ -6370,6 +6370,25 @@ def warn_fn():
             self.assertEqual(frameinfo.lineno - 6, warning.lineno)
             self.assertEqual(len(w), 1)
 
+    @unittest.skipIf(not TEST_NUMPY, 'NumPy not found')
+    @onlyOnCPUAndCUDA
+    @dtypesIfCUDA(torch.half, torch.float, torch.double)
+    @dtypes(torch.float, torch.double)
+    def test_nanprod(self, device, dtype):
+        x = ([[float('nan'), 1.23], [2.06, float('nan')]], [[float('nan'), float('nan')], [float('nan'), float('nan')]])
```

Comment:
This test is a good start, but it needs some additions:

- integer and complex types (if prod supports them)
- tests analogous to the prod tests in `tensor_op_tests`, which are also in test_torch.py
- tests for edgecases (e.g. an empty tensor)
- tests for tensors of dims 0, 1, and 3
- generate the input tensors using randn and then populate nans randomly 
- tests setting the dtype kwarg

Check out the test in https://github.com/pytorch/pytorch/pull/39992 for how to generate the tensors. You can enumerate over the axes, too, when the tensor is 3D. 

You may want to special case setting the dtype kwarg. Although I'm curious what will happen if you try to pass an integer dtype when the input tensor is float or complex, and a float tensor if the input tensor is complex.

Follow-up question: does (nan)prod support bool inputs?

Reply:
Update: I've updated tests, not supporting complex for this PR. Also, nanprod and prod supports bool inputs.

Still working on adding edge cases, if required.